vmproxy/pkg/proxy/proxy.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2022 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package proxy provides http reverse proxy functionality
+package proxy
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Config defines options for starting proxy
+type Config struct {
+	// Name of the header to check for filters. Case insensitive.
+	HeaderName string
+	// Address the proxy is listening on
+	ListenAddress string
+	// Target URL to forward requests to
+	TargetURL *url.URL
+}
+
+// StartProxy starts proxy which adds extra filters based on configuration.
+func StartProxy(cfg Config) {
+	logrus.Infof("Starting to proxy at http://%s to %s", cfg.ListenAddress, cfg.TargetURL.String())
+
+	err := http.ListenAndServe(cfg.ListenAddress, getHandler(cfg))
+	logrus.Error(err)
+}
+
+func getHandler(cfg Config) http.HandlerFunc {
+	rProxy := &httputil.ReverseProxy{
+		Director: director(cfg.TargetURL, cfg.HeaderName),
+	}
+
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		if failOnInvalidHeader(rw, req, cfg.HeaderName) {
+			return
+		}
+
+		rProxy.ServeHTTP(rw, req)
+	})
+}
+
+func failOnInvalidHeader(rw http.ResponseWriter, req *http.Request, headerName string) bool {
+	if filters := req.Header.Get(headerName); filters != "" {
+		if _, err := parseFilters(filters); err != nil {
+			rw.Header().Set("Content-Type", "text/plain; charset=utf-8")
+			rw.WriteHeader(http.StatusPreconditionFailed)
+			io.WriteString(rw, "Failed to parse headers")
+			return true
+		}
+	}
+
+	return false
```

Comment:
I think it should fail in case header is absent in the incoming request

Reply:
The scope of this tool is to restrict access to metrics as defined in the incoming request.
In case the incoming request does not define any restrictions, the is no filtering applied.

`managed` is out of scope of this tool.
We can discuss further if this decision affects Access Control in any way. I believe it's fine and shall respect all requirements.

Reply:
> another point to think over - should `vmproxy` accept all incoming requests or only requests from `managed` (how to determine that request is sent by `managed`) ?

This is out of scope of `vmproxy`. It needs to be managed on a different level.

Reply:
Ok, we can start with it now. Later, depending on future requirements, we can improve this component.